libbeat/beat/metrics.go

```diff
@@ -0,0 +1,33 @@
+package beat
+
+import (
+	"runtime"
+
+	"github.com/elastic/beats/libbeat/monitoring"
+)
+
+type memstatsVar struct{}
+
+var (
+	metrics = monitoring.Default.NewRegistry("beat")
+)
+
+func init() {
+
+	var ms memstatsVar
+	metrics.Add("memstats", ms, monitoring.Reported)
+}
+
+func (memstatsVar) Visit(m monitoring.Mode, V monitoring.Visitor) {
+	var stats runtime.MemStats
+	runtime.ReadMemStats(&stats)
+
+	V.OnRegistryStart()
+	defer V.OnRegistryFinished()
+
+	monitoring.ReportInt(V, "memory.total", int64(stats.TotalAlloc))
```

Comment:
when using 'ReportXXX', the names should not include `.`, as `.` is used to split namespaces, but having the collector do the splitting makes things quite more complicated in the collector. Instead namespaces should be reported only once (without `.`).

For example:

```
type memstatsVar struct{}

var metrics = monitoring.Default.NewRegistry("beat")

func init() {
	metrics.Add("memstats", memstatsVar{}, monitoring.Reported)
}

func (memstatsVar) Visit(m monitoring.Mode, V monitoring.Visitor) {
	var stats runtime.MemStats
	runtime.ReadMemStats(&stats)

	V.OnRegistryStart()
	defer V.OnRegistryFinished()

	reportNamespace(V, "memory", func() {
		monitoring.ReportInt(V, "total", int64(stats.TotalAlloc))
		if m == monitoring.Full {
			monitoring.ReportInt(V, "alloc", int64(stats.Alloc))
		}
	})

	if m == monitoring.Full {
		reportNamespace(V, "gc", func() {
			monitoring.ReportInt(V, "next", int64(stats.NextGC))
		})
	}
}

func reportNamespace(V monitoring.Visitor, name string, f func()) {
	V.OnKey(name)
	V.OnRegistryStart()
	defer V.OnRegistryFinished()
	f()
}
```